src/protocols/rdp/argv.c

```diff
@@ -0,0 +1,192 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "config.h"
+#include "argv.h"
+#include "rdp.h"
+
+#include <guacamole/protocol.h>
+#include <guacamole/socket.h>
+#include <guacamole/user.h>
+
+#include <pthread.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * All RDP connection settings which may be updated by unprivileged users
+ * through "argv" streams.
+ */
+typedef enum guac_rdp_argv_setting {
+
+    /**
+     * The username of the connection.
+     */
+    GUAC_RDP_ARGV_SETTING_USERNAME,
+            
+    /**
+     * The password to authenticate the connection.
+     */
+    GUAC_RDP_ARGV_SETTING_PASSWORD,
+            
+    /**
+     * The domain to use for connection authentication.
+     */
+    GUAC_RDP_ARGV_SETTING_DOMAIN,
+    
+} guac_rdp_argv_setting;
+
+/**
+ * The value or current status of a connection parameter received over an
+ * "argv" stream.
+ */
+typedef struct guac_rdp_argv {
+
+    /**
+     * The specific setting being updated.
+     */
+    guac_rdp_argv_setting setting;
+
+    /**
+     * Buffer space for containing the received argument value.
+     */
+    char buffer[GUAC_RDP_ARGV_MAX_LENGTH];
+
+    /**
+     * The number of bytes received so far.
+     */
+    int length;
+
+} guac_rdp_argv;
+
+/**
+ * Handler for "blob" instructions which appends the data from received blobs
+ * to the end of the in-progress argument value buffer.
+ *
+ * @see guac_user_blob_handler
+ */
+static int guac_rdp_argv_blob_handler(guac_user* user,
+        guac_stream* stream, void* data, int length) {
+
+    guac_rdp_argv* argv = (guac_rdp_argv*) stream->data;
+    
+    /* Calculate buffer size remaining, including space for null terminator,
+     * adjusting received length accordingly */
+    int remaining = sizeof(argv->buffer) - argv->length - 1;
+    if (length > remaining)
+        length = remaining;
+
+    /* Append received data to end of buffer */
+    memcpy(argv->buffer + argv->length, data, length);
+    argv->length += length;
+
+    return 0;
+
+}
+
+/**
+ * Handler for "end" instructions which applies the changes specified by the
+ * argument value buffer associated with the stream.
+ *
+ * @see guac_user_end_handler
+ */
+static int guac_rdp_argv_end_handler(guac_user* user,
+        guac_stream* stream) {
+
+    guac_client* client = user->client;
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    guac_rdp_settings* settings = rdp_client->settings;
+    
+    /* Append null terminator to value */
+    guac_rdp_argv* argv = (guac_rdp_argv*) stream->data;
+    argv->buffer[argv->length] = '\0';
+
+    /* Apply changes to chosen setting */
+    switch (argv->setting) {
+
+        /* Update RDP username. */
+        case GUAC_RDP_ARGV_SETTING_USERNAME:
+            free(settings->username);
+            settings->username = malloc(strlen(argv->buffer) * sizeof(char));
+            strcpy(settings->username, argv->buffer);
```

Comment:
`strdup()` could be easier than `malloc(strlen(...))` followed by `strcpy()`. That said, it looks like already have the length with `argv->length`, and so manually terminating the input, recalculating the length with `strlen()`, etc. feels superfluous.

Perhaps `strndup(..., argv->length)` would solve everything?

Reply:
Less code, I like it.  Fixed.